0006/README.md

```diff
@@ -0,0 +1,318 @@
+    [version branches](#version-branches). As a result, `.gitmodules` is added
+    in the top level directory, and has `url = ./` for each submodule config to
+    ensure referencing to the local repo. A future version should follow a
+    similar setup in the repo.
+-   All permalinks in file content use the dir structure described in
+    [the website](#the-website) section.
+
+NOTE: We admit the way of using the Git submodule is non-standard, but it is a
+nice compromise and helps us to improve the maintenance experience.
+
+#### Version Branches
+
+Git branch is the way to manage versioned specifications.
+
+*What is changing*
+
+-   `release/<version|version-draft>` is the branch pattern for a prior or
+    future `version`, where the `version` must contain both major and minor.
+-   In addition to the shared images in the `main` branch, if an image is
+    required for a specific version, the `/images` directory in the version
+    branch may contain the image and let the versioned files reference it .
+-   At the top level directory, a branch has all the versioned file from
+    `/docs/spec/<version>/*`, `/docs/provenance/<version>.md`,
+    `/docs/verification_summary/<version>.md`, and
+    `/docs/github-actions-workflow/<version>.md`. All specification files in the
+    same branch share the same version.
+
+#### Code Contribution
+
+Since the repo uses the Git submodule in a non-standard way and for easy
+building website purposes, people SHOULD NOT use submodules to commit any
+change.
+
+When a community member needs to make a contribution, they fork the repo (and
+clone it locally if they are familiar with git). If the change is for a
+versioned file (a.k.a. spec file), they should make the change in the version
+branch, and send a PR to the version branch. Meanwhile, if the change is for an
+unversioned file or the website itself, they should make the change in `main`,
+and send a PR to `main`.
+
+However, if the change comes from one of the repo owners, and it has to modify
+files in both `main` and version branches in one PR, we assume the owner
+understands the risk and knows how to send a proper code update with Submodule
+changes.
+
+#### Submodule Pointer
+
+A new change in a version branch makes the Submodule pointer in the `main`
+branch stale. A Github Workflow script may be created to automatically update
+the Submodule pointing to the latest commit, when the PR is merged. If desired,
+the repo owner can also manually launch the Github Workflow.
+
+## Automate Building and Deploying The Website
+
+A shell script should assemble the website, instead of using any logic in the
+Github Workflow script.
+
+### Checkout repo, load all submodules, and build once
+
+The GitHub Workflow clones the `main` branch once, pulls the latest files for
+all the submodules, builds the website once, and deploys the website to Github
+Pages.
+
+*Disadvantages*
+
+-   The usage of Git submodule is non-standard, and it leaves some versioned
```

Comment:
What sorts of precautions do we need to take to make sure when using submodules to ensure we don't end up with cyclic dependencies?

Correct me if I'm wrong, but let's say for example we create SLSA v2-rc while still updating SLSA v1.5. If we have submodules pointing to those branches and if those branches themselves still contain the submodules you could end up with a situation where v2-rc points to v1.5 as a submodule and v1.5 points back to v2-rc right?

I would just suggest we enforce a fixed depth. Also if possible we avoid the branches themselves having references submodules and ensure only main has the references.

Reply:
Let me copy part of the example above from the "The Repo" section.

```
slsa-framework/slsa:main
├─ docs/
│  ├─ ...
│  ├─ v0.1/    # a submodule points to the branch `release/v0.1`.
│  ├─ v1.0/    # a submodule points to the branch `release/v1.0`.
│  ├─ ...
│  └─ <other-unversioned-files>.md
├─ README.md
└─ ...

slsa-framework/slsa:release/v1.0 ===> ASSUME THIS IS  `SLSA v2-rc` from YOUR EXAMPLE
├─ ...
└─ <v1.0-versioned-files>.md

slsa-framework/slsa:release/v0.1 ===> ASSUME THIS IS  `SLSA v1.5` from YOUR EXAMPLE
├─ ...
└─ <v0.1-versioned-files>.md
```

Every version branch has a submodule that exists exactly once in main. Updating `SLSA v2-rc` and `SLSA v1.5` has nothing to do with submodules. Both version branches are self-contained and should only have spec files, and maybe some images (If a PR tries to violate the rule, maybe the repo owner should reject it?). The updated "Code Contribution" section explicitly mentioned that submodules should not be used to commit any changes. With that said, the depth is going to be one, and no submodule exist outside the main branch. However, please let me know if I missed any corner case here.

Mark and I will meet again to discuss the other subtree solution. You may see an update coming soon.

Reply:
Updated the content to propose using Git subtrees.